### ISSUE TITLE
Linkify/React does not generate keys correctly, resulting in lost content

### ISSUE THREAD

**Issue by gdehmlow** [SoapBox/linkifyjs#217]
## Problem

```jsx
    <Linkify>
      <div key="key-0">
        <div key="key-0">Text one</div>
        <div key="key-1">
          <div key="key-0">Text two</div>
        </div>
        <div key="key-2">Text three</div>
      </div>
    </Linkify>
```

This will only display:

```
Text one
Text two
```

And throw a warning:

```
Warning: flattenChildren(...): Encountered two children with the same key, `linkified-element-4`. Child keys must be unique; when two children share a key, only the first child will be used.
```

## Why it's happening

This is because the keys that are currently being generated look like this:

```jsx
  <div key="linkified-element-4">
    <div key="linkified-element-2">Text one</div> 
    <div key="linkified-element-4">
      <div key="linkified-element-4">Text two</div>
    </div>
    <div key="linkified-element-4">Text three</div>
  </div>
```

So the parent of `Text two` and `Text three` end up having the same key, so `Text three` is dropped.  

## How to fix

[React keys](https://reactjs.org/docs/lists-and-keys.html#keys) only need to be unique _within_ an array to uniquely identify that item in the array.

This is the way the keys are being generated:

```js
function linkifyReactElement(element, opts, elementId = 0) {
        // ...
	React.Children.forEach(element.props.children, (child) => {
		if (typeof child === 'string') {
			children.push(...stringToElements(child, opts));
		} else if (React.isValidElement(child)) {
			if (typeof child.type === 'string'
				&& options.contains(opts.ignoreTags, child.type.toUpperCase())
			) {
				children.push(child);
			} else {
				children.push(linkifyReactElement(child, opts, ++elementId)); // We increment here
			}
		} else {
			children.push(child);
		}
	});

	let newProps = {key: `linkified-element-${elementId}`}; // And then use the incremented value here
```

Notice that we're incrementing the `elementId` within the child loop.. and then using that incremented value in the parent.

What we should be doing instead is either: using an instance variable within the render function and moving `linkifyReactElement` inside of the `Linkify` component, the way that [react-linkify does it](https://github.com/tasti/react-linkify/blob/master/src/Linkify.jsx#L74) with its `parseCounter`, which results in this:

<img width="264" alt="screen shot 2017-11-16 at 11 52 10 am" src="https://user-images.githubusercontent.com/850025/32912668-9e3ac96c-cac4-11e7-93f9-897212a1a9fe.png">

Or you could just do `let childElementId = 0;` before looping over the children, and using that instead.. because only _siblings_ in arrays need unique keys.

**Comment by tone81** [comment]
@gdehmlow, this issue is related to my attempt to fix the warning in https://github.com/SoapBox/linkifyjs/pull/215.

**Comment by nfrasser** [comment]
Fixed in #215
